src/extension/infer.rs

```diff
@@ -1091,4 +1042,76 @@ mod test {
         }
         Ok(())
     }
+
+    #[test]
+    fn extension_adding_sequence() -> Result<(), Box<dyn Error>> {
+        fn mknode(
+            hugr: &mut Hugr,
+            root: Node,
+            sig: FunctionType,
+            ext: ExtensionId,
+        ) -> Result<Node, Box<dyn Error>> {
+            let [node, input, output] = create_with_io(
+                hugr,
+                root,
+                ops::DFG {
+                    signature: sig
+                        .clone()
+                        .with_extension_delta(&ExtensionSet::singleton(&ext)),
+                },
+            )?;
+
+            let lift = hugr.add_node_with_parent(
+                node,
+                NodeType::open_extensions(ops::LeafOp::Lift {
+                    type_row: type_row![NAT],
+                    new_extension: ext,
+                }),
+            )?;
+
+            hugr.connect(input, 0, lift, 0)?;
+            hugr.connect(lift, 0, output, 0)?;
+
+            Ok(node)
+        }
+
+        let df_sig = FunctionType::new(type_row![NAT], type_row![NAT]);
+
+        let mut hugr = Hugr::new(NodeType::open_extensions(ops::DFG {
+            signature: df_sig
+                .clone()
+                .with_extension_delta(&ExtensionSet::from_iter([A, B])),
+        }));
+
+        let root = hugr.root();
+        let input = hugr.add_node_with_parent(
+            root,
+            NodeType::open_extensions(ops::Input {
+                types: type_row![NAT],
+            }),
+        )?;
+        let output = hugr.add_node_with_parent(
+            root,
+            NodeType::open_extensions(ops::Output {
+                types: type_row![NAT],
+            }),
+        )?;
+
+        let node0 = mknode(&mut hugr, root, df_sig.clone(), A)?;
+        let node1 = mknode(&mut hugr, root, df_sig.clone(), A)?;
+        let node2 = mknode(&mut hugr, root, df_sig.clone(), B)?;
+        let node3 = mknode(&mut hugr, root, df_sig.clone(), B)?;
+        let node4 = mknode(&mut hugr, root, df_sig.clone(), A)?;
+        let node5 = mknode(&mut hugr, root, df_sig.clone(), B)?;
+
+        // Connect nodes in order (0 -> 1 -> 2 ...)
+        let nodes = [input, node0, node1, node2, node3, node4, node5, output];
```

Comment:
How about something like
```
let mid_nodes = [A,A,B,B,A,B].map(|ext| mknode(&mut hugr, root, df_sig.clone(), ext)).collect::<Result<Vec<Node>,_>>()?;
let nodes = [[input], mid_nodes, [output]].concat()
```
(there is a collect that turns iterator of Result<T,...> into Result<Vec<T>,...>)

Reply:
You could even define `mknode` as a closure `|ext| {...}` after you've done `let hugr =` and then it'd be just `.map(mknode).collect....`